### PR TITLE
Global Energy Functions

### PR DESCRIPTION
**Pull request by purujitgoyal**

### CODE REVIEW

File: structured_prediction_baselines/models/base.py [hunk 1]
```diff
@@ -92,7 +92,10 @@ def get_extra_args_for_loss(
         labels: torch.Tensor,
         **kwargs: Any,
     ) -> Dict:
-        return {}
+        mask = util.get_text_field_mask(x)
```
Comment (dhruvdcoder) [inline structured_prediction_baselines/models/base.py:95]: Calculation of the mask should not be in the base model. This will break the multilabel model! This function should be overridden in the sequence tagging model only.

File: structured_prediction_baselines/modules/sampler/sequence_tagging/sequence_tagging_inference.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+        if isinstance(x, torch.Tensor):
+            return x.shape[0]
+        else:
+            raise NotImplementedError
+
+    @contextlib.contextmanager
+    def param_grad(self) -> Generator[None, None, None]:
+        trainable_modules = [self.inference_nn, self.cost_augmented_layer]
+        requires_grad_map = {}
+        for module in trainable_modules:
```
Comment (dhruvdcoder) [inline structured_prediction_baselines/modules/sampler/sequence_tagging/sequence_tagging_inference.py:134]: I think this will not work. This for loop should be repeated both in try and finally.

File: structured_prediction_baselines/modules/sampler/sequence_tagging/sequence_tagging_inference.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+            if module is not None:
+                try:
+                    for param in module.parameters():
+                        param.requires_grad = True
+                    yield
+                finally:
+                    # set the requires_grad back to false
```
Comment (dhruvdcoder) [inline structured_prediction_baselines/modules/sampler/sequence_tagging/sequence_tagging_inference.py:141]: You will need to re-loop through the trainable_modules here again.

File: structured_prediction_baselines/modules/sampler/sequence_tagging/sequence_tagging_inference.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+                    for param in module.parameters():
+                        param.requires_grad = False
+            else:  # if there is no loss_fn, we have nothing to do.
+                warnings.warn(
+                    (
+                        "There is inference_nn in gradient based inference sampler."
```
Comment (dhruvdcoder) [inline structured_prediction_baselines/modules/sampler/sequence_tagging/sequence_tagging_inference.py:147]: Change this warning to a more helpful one like: "You are using inference network-based model but cost_augmented_layer and interence_nn are both None. Either you are not using the right Sampler or you have not constructed it correctly."

File: structured_prediction_baselines/modules/sampler/sequence_tagging/sequence_tagging_inference.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+            if module is not None:
+                try:
+                    for param in module.parameters():
+                        param.requires_grad = True
+                    yield
+                finally:
+                    # set the requires_grad back to false
+                    for param in module.parameters():
+                        param.requires_grad = False
+            else:  # if there is no loss_fn, we have nothing to do.
```
Comment (dhruvdcoder) [inline structured_prediction_baselines/modules/sampler/sequence_tagging/sequence_tagging_inference.py:144]: Drop the irrelevant comment "# if there is no ...". It will confuse people.

File: structured_prediction_baselines/modules/sampler/sequence_tagging/sequence_tagging_inference.py [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+        y_hat: torch.Tensor,
+        y_cost_aug: torch.Tensor,
+        buffer: Dict,
+        loss_fn: Callable[[torch.Tensor, torch.Tensor, Dict], torch.Tensor],
+    ) -> torch.Tensor:
+        """
+        Returns:
+            loss: loss value at the previous point (unreduced)
+        """
+        loss = loss_fn(y_hat, y_cost_aug, buffer)
```
Comment (dhruvdcoder) [inline structured_prediction_baselines/modules/sampler/sequence_tagging/sequence_tagging_inference.py:219]: Before you compute loss, you need to call `optimizer.zero_grad(set_to_none=True)` to zero out the grads. Ref: https://pytorch.org/docs/stable/optim.html#optimizer-step

Comment (dhruvdcoder) [reply structured_prediction_baselines/modules/sampler/sequence_tagging/sequence_tagging_inference.py:219]: using set_to_none=True is faster than zero_grad(False).

Comment (dhruvdcoder) [reply structured_prediction_baselines/modules/sampler/sequence_tagging/sequence_tagging_inference.py:219]: In fact, I think you should call zero_grad before any forward steps are done. So perhaps on line 190.

File: structured_prediction_baselines/modules/sampler/sequence_tagging/sequence_tagging_inference.py [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+            step_number = 0
+            loss_value: Union[torch.Tensor, float] = float("inf")
+
+            # we need to enable grad because if the top-level model
+            # was being called in a validation loop, the training
+            # flag will be False for all modules. This will not allow
+            # gradient based inference to progress.
+            with torch.enable_grad():
+                while not self.stopping_criteria(step_number, float(loss_value)):
+                    y_inf: torch.Tensor = self.inference_nn(x, buffer)  # (batch_size, ...)
```
Comment (dhruvdcoder) [inline structured_prediction_baselines/modules/sampler/sequence_tagging/sequence_tagging_inference.py:190]: You should call `optimizer.zero_grad(set_to_none=True)` here before computing y_inf, instead of calling it in update().

**Review by dhruvdcoder — CHANGES_REQUESTED** [review]
Overall, the structure looks good. There are some minor issues that I have stated below. Please address these and try to run at least one training of sequence tagging model and re-run the multilabel classification model before merging. So that we are sure everything is working.

File: structured_prediction_baselines/models/sequence_tagging_bert.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+from overrides import overrides
+from torch.nn.functional import one_hot
+
+from structured_prediction_baselines.modules.score_nn import ScoreNN
+from structured_prediction_baselines.modules.structured_energy.structured_energy import StructuredEnergy
+from structured_prediction_baselines.modules.task_nn import TaskNN
+
+
+@Model.register('sequence-tagging-bert')
+class SequenceTaggingBert(Model):
```
[the remark this answers is not in the document]
Comment (purujitgoyal) [reply structured_prediction_baselines/models/sequence_tagging_bert.py:23]: Ignore this file. This was the one I created first before everything else was added. The correct model is here:
https://github.com/dhruvdcoder/structured_prediction_baselines/blob/energy_functions/structured_prediction_baselines/models/sequence_tagging.py
